realm/realm-library/src/objectServer/java/io/realm/internal/events/ChangeEvent.java

```diff
@@ -0,0 +1,234 @@
+
+        final BsonDocument nsDoc = new BsonDocument();
+        nsDoc.put(Fields.NS_DB_FIELD, new BsonString(ns.getDatabaseName()));
+        nsDoc.put(Fields.NS_COLL_FIELD, new BsonString(getNamespace().getCollectionName()));
+        asDoc.put(Fields.NS_FIELD, nsDoc);
+
+        asDoc.put(Fields.DOCUMENT_KEY_FIELD, getDocumentKey());
+
+        if (getFullDocument() != null && (getFullDocument() instanceof BsonValue)
+                && ((BsonValue) getFullDocument()).isDocument()) {
+            asDoc.put(Fields.FULL_DOCUMENT_FIELD, (BsonValue) getFullDocument());
+        }
+
+        if (getUpdateDescription() != null) {
+            asDoc.put(Fields.UPDATE_DESCRIPTION_FIELD, getUpdateDescription().toBsonDocument());
+        }
+
+        asDoc.put(Fields.WRITE_PENDING_FIELD, new BsonBoolean(hasUncommittedWrites()));
+        return asDoc;
+    }
+
+    /**
+     * Deserializes a {@link BsonDocument} into an instance of change event.
+     * @param document the serialized document
+     * @return the deserialized change event
+     */
+    static <T> ChangeEvent<T> fromBsonDocument(final BsonDocument document, final Class<T> documentClass, CodecRegistry codecRegistry) {
+        keyPresent(Fields.ID_FIELD, document);
+        keyPresent(Fields.OPERATION_TYPE_FIELD, document);
+        keyPresent(Fields.NS_FIELD, document);
+        keyPresent(Fields.DOCUMENT_KEY_FIELD, document);
+
+        final BsonDocument nsDoc = document.getDocument(Fields.NS_FIELD);
+
+        final UpdateDescription updateDescription;
+        if (document.containsKey(Fields.UPDATE_DESCRIPTION_FIELD)) {
+            updateDescription = UpdateDescription.fromBsonDocument(
+                    document.getDocument(Fields.UPDATE_DESCRIPTION_FIELD)
+            );
+        } else {
+            updateDescription = null;
+        }
+
+        final T fullDocument;
+
+        if (document.containsKey(Fields.FULL_DOCUMENT_FIELD)) {
+            final BsonValue fdVal = document.get(Fields.FULL_DOCUMENT_FIELD);
+            if (fdVal.isDocument()) {
+                fullDocument = codecRegistry.get(documentClass).decode(fdVal.asDocument().asBsonReader(), DecoderContext.builder().build());
+            } else {
+                fullDocument = null;
+            }
+        } else {
+            fullDocument = null;
+        }
+
+        return new ChangeEvent<>(
+                document.getDocument(Fields.ID_FIELD),
+                fromRemote(document.getString(Fields.OPERATION_TYPE_FIELD).getValue()),
+                fullDocument,
+                new MongoNamespace(
+                        nsDoc.getString(Fields.NS_DB_FIELD).getValue(),
+                        nsDoc.getString(Fields.NS_COLL_FIELD).getValue()),
+                document.getDocument(Fields.DOCUMENT_KEY_FIELD),
+                updateDescription,
+                document.getBoolean(Fields.WRITE_PENDING_FIELD, BsonBoolean.FALSE).getValue());
+    }
+
+    private static final class Fields {
```

Comment:
There are a lot of magic constants here. Is there a spec or something we can refer to for the source of truth for these?

Reply:
This hasn't been addressed.